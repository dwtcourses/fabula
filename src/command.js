
export default class Command {
  constructor(cmd, line, match) {
    this.cmd = { ...cmd }
    this.params = {}
    this.match = match
    this.argv = this._expandTildes(line.split(/\s+/))
    this.source = []
    this.firstLine = true
  }
  _expandTildes(argv) {
    return argv.map((arg) => {
      if (arg.startsWith('~')) {
        return `${process.env.HOME}${arg.slice(1)}`
      }
      return arg
    })
  }
  handleLine(line) {
    if (!this.cmd.line) {
      this.source.push(line)
      return false
    }
    if (this.firstLine) {
      this.firstLine = false
    }
    const continueCommand = this.cmd.line.call(this, line)
    this.source.push(line)
    return continueCommand
  }
  run() {
    this.cmd.command.apply(this)
  }
}
