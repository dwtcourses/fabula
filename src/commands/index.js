
export default [
  () => import('./fabula'),
  () => import('./ensure'),
  () => import('./get'),
  () => import('./put'),
  () => import('./write'),
  () => import('./cd')
]
