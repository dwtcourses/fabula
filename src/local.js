import { writeFileSync, appendFileSync } from 'fs'
import { spawn } from 'child_process'

export function execLocal(cmd, env = {}, cwd = null) {
  return new Promise((resolve) => {
    let stdout = ''
    let stderr = ''
    const options = { env }
    if (cwd) {
      options.cwd = cwd
    }
    const stream = spawn(...cmd, options)
    stream.on('close', (code) => {
      resolve({ stdout, stderr, code })
    })
    stream.on('data', (data) => { stdout += data })
    stream.stderr.on('data', (data) => { stderr += data })
  })
}

export function localWrite(filePath, fileContents) {
  writeFileSync(filePath, fileContents)
}

export function localAppend(filePath, fileContents) {
  appendFileSync(filePath, fileContents)
}
