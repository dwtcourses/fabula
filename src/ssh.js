
import readline from 'readline'
import { Writable } from 'stream'
import { readFileSync } from 'fs'
import { promisify } from 'util'
import { Client } from 'ssh2'

function askPassphrase(privateKey) {
  return new Promise((resolve) => {
    const passphraseOutput = new Writable({
      write(chunk, encoding, callback) {
        if (!this.protected) {
          process.stdout.write(chunk, encoding)
        } else {
          // Displays typed passphrase as * on the screen
          const withAsterisks = chunk.toString().replace(/./g, '*')
          process.stdout.write(Buffer.from(withAsterisks), encoding)
        }
        callback()
      }
    })
    const rl = readline.createInterface({
      input: process.stdin,
      output: passphraseOutput,
      terminal: true
    })
    rl.question(`${privateKey} requires a passphrase: `, (passphrase) => {
      rl.close()
      resolve(passphrase)
    })
    passphraseOutput.protected = true
  })
}

function isKeyEncrypted(privateKey) {
  if (/^-*BEGIN ENCRYPTED PRIVATE KEY/g.test(privateKey)) {
    return true
  } else {
    const firstLines = privateKey.split(/\n/g).slice(0, 2)
    return firstLines.some((line) => line.match(/Proc-Type: 4,ENCRYPTED/))
  }
}

export function getConnection(settings) {
  return new Promise(async (resolve, reject) => {
    const conn = new Client()
    conn.exec = promisify(conn.exec).bind(conn)
    conn.sftp = promisify(conn.sftp).bind(conn)
    conn.on('error', reject)
    conn.on('ready', () => resolve(conn))

    const privateKey = readFileSync(settings.privateKey).toString()
    if (!settings.passphrase && isKeyEncrypted(privateKey)) {
      settings.passphrase = await askPassphrase(settings.privateKey)
    }
    conn.connect({ ...settings, privateKey })
  })
}

export async function echo(conn, { filePath, fileContents }) {
  const stream = await conn.sftp()
  return stream.writeFile(filePath, fileContents)
}

export async function append(conn, { filePath, fileContents }) {
  // escape string
  // run echo escaped(fileContents) >> filePath on the server
}

export async function put(conn, cmd) {
  const stream = await conn.sftp()
  return stream.fastPut(cmd)
}

export function exec(conn, cmd) {
  return new Promise(async (resolve, reject) => {
    let stdout = ''
    let stderr = ''
    const stream = await conn.exec(cmd).catch(reject)
    stream.on('close', (code, signal) => {
      resolve({ stdout, stderr, code, signal })
    })
    stream.on('data', (data) => { stdout += data })
    stream.stderr.on('data', (data) => { stderr += data })
  })
}
