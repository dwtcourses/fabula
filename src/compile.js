
import Module from 'module'

import template from 'lodash.template'
import merge from 'lodash.merge'

import Command from './command'
import execCommand from './commands/exec'
import commands from './commands'
import { quote } from './utils'

function requireFromString(code, name) {
  const m = new Module()
  // Second parameter is here is a virtual unique path
  m._compile(code, `components/${name}.js`)
  m.loaded = true
  return m.exports
}

compile.loadComponent = function (source) {
  source = source.split(/\n/g)
    .filter(line => !line.startsWith('#'))

  const fabula = []
  const script = []
  const strings = []

  let match
  let element
  let string = {}
  let prepend

  for (const line of source) {
    // eslint-disable-next-line
    if (match = line.match(/^\s*<([^\/%][^>]+)>/)) {
      element = match[1].split(/\s+/g)
      prepend = element.slice(1).join(' ')
      element = element[0]
      // eslint-disable-next-line no-cond-assign
      if (match = match[1].match(/^string\s+id="([^"]+)"/)) {
        string = { id: match[1], lines: [] }
        element = 'string'
      }
      continue
    // eslint-disable-next-line no-cond-assign
    } else if (match = line.match(/^\s*<\/([^>]+)>/)) {
      if (match[1] === 'string') {
        strings.push(string)
      }
      element = null
    }
    if (!element) {
      continue
    }
    switch (element) {
      case 'fabula':
        fabula.push(line)
        break
      case 'commands':
        script.push(line)
        break
      case 'string':
        string.lines.push(line)
        break
    }
  }
  return { fabula, script, strings, prepend }
}

compile.compileTemplate = function (cmd, settings) {
  const cmdTemplate = template(cmd, {
    imports: { quote },
    interpolate: /<%=([\s\S]+?)%>/g
  })
  return cmdTemplate(settings)
}

compile.splitMultiLines = function (source) {
  let multiline
  return source.split(/\n/g).reduce((_lines, line) => {
    if (multiline) {
      line = line.trimLeft()
      multiline = /\\\s*$/.test(line)
      const index = _lines.length ? _lines.length - 1 : 0
      _lines[index] += line.replace(/\s*\\\s*$/, ' ')
      return _lines
    }
    multiline = /\\\s*$/.test(line)
    return _lines.concat([line.replace(/\s*\\\s*$/, ' ')])
  }, [])
}

compile.parseLine = function (commands, command, line, prepend, settings, env, push) {
  let cmd
  if (command) {
    if (command.handleLine(line)) {
      return command
    } else {
      command = null
    }
  }
  let match
  const commandSearchList = [...commands]
  if (settings.commands) {
    commandSearchList.push(...settings.commands)
    delete settings.commands
  }
  commandSearchList.push(execCommand)
  let _line
  for (cmd of commandSearchList) {
    if (cmd.match) {
      command = new Command(cmd, line, env)
      command.settings = settings
      _line = command.registerHandler(line)
      if (prepend && !/^\s+/.test(_line)) {
        _line = command.prepend(prepend, _line)
      }
      match = command.cmd.match.call(command, _line)
      if (match) {
        break
      }
    }
  }
  if (match) {
    command.match = match
  }
  if (command.handleLine(_line)) {
    push(command)
    return command
  } else {
    push(command)
  }
}

function compileComponent(name, source, settings) {
  const { fabula, script, strings, prepend } = compile.loadComponent(source)
  const componentSource = script.join('\n')

  const _componentSettings = requireFromString(fabula.join('\n'))
  const componentSettings = _componentSettings.default || _componentSettings

  const globalEnv = { ...settings.env }
  delete globalEnv.local
  delete globalEnv.ssh

  const env = {
    ...globalEnv,
    ...componentSettings.env
  }
  delete componentSettings.env

  // Can't override any of these from a component
  delete componentSettings.agent
  delete componentSettings.ssh

  merge(settings, componentSettings)

  settings.strings = strings.reduce((hash, string) => {
    const compiledString = compile.compileTemplate(string.lines.join('\n'), settings)
    return { ...hash, [string.id]: quote(compiledString, true) }
  }, {})

  return compile(name, componentSource, settings, prepend, env)
}

export async function compile(name, source, settings, prepend, env = {}) {
  // If <fabula> or at least <commands> is detected,
  // process as a component and return
  if (source.match(/^\s*<(?:(?:fabula)|(commands))[^>]*>/g)) {
    return compileComponent(name, source, settings)
  }

  const _vars = {}
  settings.vars = new Proxy(_vars, {
    get(obj, prop) {
      if (prop in obj) {
        return obj[prop]
      } else {
        return settings.strings[prop]
      }
    }
  })

  // If no marked section is found, proceed to
  // regular commands compilation
  source = compile.compileTemplate(source, settings)

  // splitMultiLines() will merge lines ending in `\`
  // with the subsequent one, preserving Bash's behaviour
  const lines = compile.splitMultiLines(source)
    .filter(Boolean)
    .filter(line => !line.startsWith('#'))

  const _commands = await Promise.all(
    commands.map(cmd => cmd().then(c => c.default))
  )

  let currentCommand
  const parsedCommands = []

  for (const line of lines) {
    // If a component's line() handler returns true,
    // the same command object will be returned, allowing
    // parsing of custom mult-line special commands
    currentCommand = compile.parseLine(_commands, currentCommand, line, prepend, settings, env, (command) => {
      parsedCommands.push(command)
    })
  }

  return parsedCommands
}
