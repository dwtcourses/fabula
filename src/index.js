
import { readFileSync } from 'fs'
import template from 'lodash.template'
import consola from 'consola'

import { getConnection } from './ssh'

import Command from './command'
import commands from './commands'
import execCommand from './commands/exec'

compile.compileTemplate = function(cmd, settings) {
  const cmdTemplate = template(cmd, {
    interpolate: /<%=([\s\S]+?)%>/g
  })
  return cmdTemplate(settings)
}

compile.matchCommand = function(line, next) {
  let command
  for (const cmd of commands) {
    let match
    if (cmd.match) {
      match = cmd.match(line)
    }
    if (match) {
      return new Command(cmd, match, line, next)
    } else {
      console.log('return execCommand', line)
      return new Command(execCommand, match, line, next)  
    }
  }
}

export function compile(source, settings) {
  source = compile.compileTemplate(source, settings)

  const lines = source.split(/\n/g)
    .filter(Boolean)
    .filter(line => !line.startsWith('#'))

  let currentCommand
  const parsedCommands = []

  for (const line of lines) {
    currentCommand = compile.matchCommand(line, () => {
      currentCommand = null
    })
    if (currentCommand) {
      parsedCommands.push(currentCommand)
    }
  }
  return parsedCommands
}

// export async function run(config, task) {
//   const base = config.srcDir
//   const servers = (config.ops || {}).servers || {}
//   task = readFileSync(base, 'tasks', task)

//   for (const server in servers) {
//     const conn = getConnection(server, servers[server])
//     const template = compileTemplate(task, servers[server])
//     const tree = compileTree(template)
//     const commands = commandsFromTree(tree)
//     for (const command of commands) {
//       consola.info('Running command:', command.meta)
//       await command()
//     }
//   }
//   if (servers.length === 0) {
//     consola.warn('No servers configured.')
//   }
// }

// Mostly temporary, for testing
export async function runString(settings, str) {
  const commands = compile(str, settings)
  for (const command of commands) {
    consola.info('Running command:', command.source[0], command.params)
    // await command()
  }
}
