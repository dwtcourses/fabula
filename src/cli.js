
import { existsSync } from 'fs'
import { resolve } from 'path'
import consola from 'consola'
import arg from 'arg'
import { run } from './compile'

function resolvePath(path) {
  return resolve(process.cwd(), ...path.split('/'))
}

function loadConfig() {
  let rcFile
  for (rcFile of ['fabula.js', '.fabularc.js', '.fabularc']) {
    if (existsSync(resolvePath(rcFile))) {
      break
    }
    rcFile = null
  }
  if (rcFile === null) {
    consola.fatal('Fabula configuration file not found.')
    process.exit()
  }
  return require(resolvePath(rcFile)).default
}

function showHelpAndExit() {
  process.stdout.write(
    '\n' +
    '  Usage: fabula <server1,server2,...> <task> (run on specified servers)\n' + 
    '         fabula all <task> (run on all servers)\n' + 
    '         fabula <task> (run local only)\n\n'
  )
  process.exit()
}

export default async function () {
  const args = arg({})
  if (args._.length == 0 || args._[0] === 'help') {
    showHelpAndExit()
  }
  const config = loadConfig()
  if (args._.length === 2) {
    // Run on remote servers:
    // fabula <server1,server2,..> <script>
    // fabula all <script>
    const servers = args._[0].split(/,/g)
    const source = args._[1]
    await run(source, config, servers)
  } else if (args._.length === 1) {
    // Run strictly locally (non-local commands will cause an error)
    // fabula <local-script>
    const source = args._[0]
    if (!existsSync(source)) {
      consola.fatal('Task source doesn\'t exist.')
      process.exit()
    }
    await run(source, config)
  } else {
    showHelpAndExit()
  }
}
