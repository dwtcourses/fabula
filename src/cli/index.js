
import { resolve } from 'path'
import arg from 'arg'
import { runSource } from '../compile'

function resolvePath(path, base = null) {
  if (!base) {
    base = process.cwd()
  }
  return resolve(base, ...path.split('/'))
}

function loadServers() {
  for (const rcFile in ['.fabularc', '.fabularc.js', 'fabula.js']) {
    if (existsSync(rcFile))
  }
}

void function main() {
  const args = arg()
  const source = args._[0]


}()
