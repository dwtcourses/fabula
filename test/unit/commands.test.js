
import { readFileSync } from 'fs'
import { runEcho } from '../../src/ssh'

describe('task commands', () => {
//  test('local command', () => {
//
//  })
//
//  test('remote command', () => {
//
//  })
//
//  test('put command', () => {
//
//  })
  test('echo command', () => {
    const cmd = [
      '$HOME/.ssh/config', [
        '  Host 192.168.100.100',
        '    Hostname my-server',
        '    Port 22',
        '    IdentityFile $HOME/.keys/deploy_key',
        '    User username',
        '    StrictHostKeyChecking no'
      ]
    ]
    const parsed = runEcho.parseCommand(cmd)
    expect(JSON.stringify(parsed, null, 2)).toMatchSnapshot()
  })
})
