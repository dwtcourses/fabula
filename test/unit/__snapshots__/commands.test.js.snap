// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task commands echo command 1`] = `
"{
  \\"filePath\\": \\"$HOME/.ssh/config\\",
  \\"fileContents\\": \\"Host 192.168.100.100\\\\n  Hostname my-server\\\\n  Port 22\\\\n  IdentityFile $HOME/.keys/deploy_key\\\\n  User username\\\\n  StrictHostKeyChecking no\\"
}"
`;
